probe/ebpf/flow.c

```diff
@@ -331,30 +340,53 @@ static void fill_flow(struct __sk_buff *skb, struct flow *flow)
 SOCKET(flow_table)
 int bpf_flow_table(struct __sk_buff *skb)
 {
-	__u64 tm = bpf_ktime_get_ns();
-
-	__u32 key = START_TIME_NS;
-	__u32 *sns = bpf_map_lookup_element(&u64_config_values, &key);
-	if (sns != NULL && *sns == 0) {
-		bpf_map_update_element(&u64_config_values, &key, &tm, BPF_ANY);
-	}
-
-	struct flow flow = {}, *prev;
-	fill_flow(skb, &flow);
-
-	prev = bpf_map_lookup_element(&flow_table, &flow.key);
-	if (prev) {
-		update_metrics(skb, prev, tm,
-			flow.link_layer._hash_src == prev->link_layer._hash_src);
-		__sync_fetch_and_add(&prev->last, tm - prev->last);
-	} else {
-		update_metrics(skb, &flow, tm, 1);
-		__sync_fetch_and_add(&flow.start, tm);
-		__sync_fetch_and_add(&flow.last, tm);
-
-		bpf_map_update_element(&flow_table, &flow.key, &flow, BPF_ANY);
-	}
-
-	return 0;
+    __u64 tm = bpf_ktime_get_ns();
+
+    __u32 key = START_TIME_NS;
+    __u32 *sns = bpf_map_lookup_element(&u64_config_values, &key);
+    if (sns != NULL && *sns == 0) {
+    	bpf_map_update_element(&u64_config_values, &key, &tm, BPF_ANY);
+    }
+
+    struct flow flow = {}, *prev;
+    fill_flow(skb, &flow,tm);
+
+    prev = bpf_map_lookup_element(&flow_table, &flow.key);
+    if (prev) {
+        update_metrics(skb, prev, tm,
+            flow.link_layer._hash_src == prev->link_layer._hash_src);
+        __sync_fetch_and_add(&prev->last, tm - prev->last);
+        if(prev->layers&flow.layers&4>0){
+            if(prev->transport_layer.port_src==flow.transport_layer.port_src){
+                if(prev->transport_layer.ab_syn==0){
+                    __sync_fetch_and_add(&prev->transport_layer.ab_syn, flow.transport_layer.ab_syn);
+                }
+                if(prev->transport_layer.ab_fin==0){
```

Comment:
flow.transport_layer.ab_fin != 0 need to be check too otherwise you are going to update each time until flow...ab_fin != 0

Reply:
done